### PR TITLE
[Testcase Refactoring]Make test_mem_tracker.py device-generic and add hw_classification labels

### PR DESCRIPTION
**Pull request by youlituao**
## Summary

All three tests in `test/distributed/_tools/test_mem_tracker.py`
(`test_accelerator_tracker_equivalence`,
`test_tracker_with_activation_checkpointing`, and
`test_tracker_attribution`) target accelerator memory tracking behavior
using `MemTracker`, `memory_stats`, and `torch.get_device_module(dev)`.
However, two of them were gated by a hardcoded CUDA/XPU whitelist
(`@unittest.skipIf(not TEST_CUDA and not TEST_XPU, ...)`), and the class
used manual device parameterization (`torch.accelerator.current_device_index()`
/ `torch.get_default_device()`) instead of `instantiate_device_type_tests`.

This PR consolidates all tests into a single `TestMemTrackerDevice`
(ACCELERATOR) class, replaces the whitelist with
`instantiate_device_type_tests(..., except_for="cpu", allow_xpu=True)`,
adds `hw_classification` labels, and changes device acquisition to
framework-injected `device` parameter.

Note: `except_for="cpu"` means CPU-only machines will not run these tests.
The original `test_tracker_attribution` could run on CPU (via
`torch.get_default_device()`), but since it tests `MemTracker` attribution
behavior (params/grads/optstates device memory tracking), keeping it in
the ACCELERATOR class is more appropriate.

## Changes

- Renamed `TestMemTracker` to `TestMemTrackerDevice` and consolidated all
  three test methods into this single ACCELERATOR class.
- Replaced `@unittest.skipIf(not TEST_CUDA and not TEST_XPU, ...)` with
  `instantiate_device_type_tests(TestMemTrackerDevice, globals(), except_for="cpu", allow_xpu=True)`
  — CPU variants are not generated, CUDA/NPU/XPU variants are automatically created.
- Changed device acquisition from `torch.accelerator.current_device_index()` /
  `torch.get_default_device()` to `torch.device(device)` (framework-injected
  via `instantiate_device_type_tests`).
- Added `hw_classification = HardwareClassification.ACCELERATOR` class attribute.
- Removed the now-unused `TEST_CUDA` / `TEST_XPU` / `unittest` imports.
- Fixed the grammatically incorrect skip message
  ("Neither CUDA or XPU is not available" → "No accelerator is available").

## Dependencies

This PR uses the following infrastructure (all already merged into main):

| Infrastructure | Source PR | Status |
|---|---|---|
| `HardwareClassification` enum | #186918 | Merged |
| `hw_classification` labeling examples | #190991 | Merged |

No dependency on unmerged PRs.

## Not changed

- All test bodies (model construction, training loop, assertions) are unchanged
  — only device acquisition mechanism and class structure were refactored.
- The numerical equivalence assertion in `test_accelerator_tracker_equivalence`
  (tracker peak vs. allocator `active_bytes` peak, `delta=0.1`) is unchanged.
  It implicitly assumes peak memory is dominated by user-visible tensors, which
  holds on CUDA. On backends whose kernels allocate transient workspace
  directly from the caching allocator without going through aten dispatch,
  this assumption may not hold (see test plan); that is a backend accounting
  behavior, not something this PR introduces or worsens.

## Test plan

- `python test/distributed/_tools/test_mem_tracker.py` on an out-of-tree
  accelerator backend (PrivateUse1, torch_npu 2.13.0+git45fbeae on
  PyTorch 2.13.0a0+gitfad7424, Ascend 910B3):
  `test_tracker_attribution_npu` passes;
  `test_tracker_with_activation_checkpointing_npu` passes;
  `test_accelerator_tracker_equivalence_npu` runs but its equivalence
  assertion does not hold on this backend (accuracy=0.384, allocator peak
  27,375,616 B vs. MemTracker 10,522,624 B). Investigation (optimizer-variant
  and warmup experiments) shows the in-test allocation delta is exactly
  21,042,688 B regardless of optimizer choice, and the untracked remainder
  (~4.2 MB per bf16 Linear(512,512) layer per iteration) is transient matmul
  workspace allocated by backend kernels directly from the caching allocator,
  which is invisible to dispatch-level memory tracking by design. This
  accuracy value is identical on PyTorch 2.10 and 2.13.0a0, confirming it
  is a stable backend behavior independent of the PyTorch version. This
  predates and is orthogonal to this change.
  (`FAILED (failures=1)`, 3 tests ran)
- `python test/distributed/_tools/test_mem_tracker.py` on CPU (no accelerator):
  No tests ran — `except_for="cpu"` prevents CPU variant generation.
  (`NO TESTS RAN`)
- `python test/distributed/_tools/test_mem_tracker.py` on CUDA (A100,
  PyTorch 2.13.0a0, CUDA 12.8):
  All 3 tests pass (`OK`, 3 tests ran).
  `test_accelerator_tracker_equivalence_cuda` passes (accuracy within
  `delta=0.1`); `test_tracker_with_activation_checkpointing_cuda` passes;
  `test_tracker_attribution_cuda` passes.
  This confirms the refactoring does not break existing CUDA behavior.

### CODE REVIEW

**Review by JamesD18 — COMMENTED** [review]
No methods in TestMemTracker qualify as GENERIC. Rather than a mixed‑type class, it targets accelerators exclusively.

**Comment by youlituao** [comment]
> No methods in TestMemTracker qualify as GENERIC. Rather than a mixed‑type class, it targets accelerators exclusively.

Thank you for the review. You are correct — `test_tracker_attribution` does not qualify as GENERIC.

Although it used `torch.get_default_device()` (which returns CPU when no accelerator is available), the test
purpose is accelerator memory tracking behavior: it uses `MemTracker` (a tool designed for caching-allocator
memory analysis), `torch.optim.Adam(foreach=True)`, and `with torch.device(dev):` — all of which are
accelerator-oriented. "Can run on CPU" ≠ "is device-agnostic framework logic".

I have updated the PR to consolidate all three methods into a single `TestMemTrackerDevice` (ACCELERATOR)
class with `instantiate_device_type_tests(..., except_for="cpu", allow_xpu=True)`, removing the GENERIC
class entirely. `test_tracker_attribution` now also takes a `device` parameter and uses `torch.device(device)`
instead of `torch.get_default_device()`.

Note: `except_for="cpu"` means CPU-only machines will not run these tests. The original `test_tracker_attribution`
could run on CPU (via `torch.get_default_device()`), but since it tests `MemTracker` attribution behavior
(params/grads/optstates device memory tracking), keeping it in the ACCELERATOR class is more appropriate.

File: test/distributed/_tools/test_mem_tracker.py [hunk 1]
```diff
@@ -28,16 +32,12 @@ def _reset_mem_stats(self, dev: torch.device):
         mod.reset_accumulated_memory_stats(dev)
         mod.reset_peak_memory_stats(dev)
 
-    @unittest.skipIf(
-        not TEST_CUDA and not TEST_XPU, "Neither CUDA or XPU is not available"
-    )
-    def test_accelerator_tracker_equivalence(
-        self,
-    ):
+    @skipMPS
```
Comment (JamesD18) [inline test/distributed/_tools/test_mem_tracker.py]: There is no need to add the `skipMPS` decorator.

Comment (youlituao) [reply test/distributed/_tools/test_mem_tracker.py]: Agreed, removed. Since `instantiate_device_type_tests` defaults to `allow_mps=False`, MPS variants are not generated at all — `@skipMPS` would have no effect. The decorator has been removed from both methods, and the `skipMPS` import has been cleaned up.

**Review by JamesD18 — APPROVED** [review]
LGTM.Thanks for addressing the review comments.

**Review by KarhouTam — APPROVED** [review]
Thanks.
cc @fffrog
